### PR TITLE
Implement E2E Test Using Reference Bundler

### LINKED ISSUE #145: [4337] E2E Test with Reference Bundler
## Context / issue

Currently, while we have E2E tests that verify that the 4337 module with the Safe contracts work with the reference entry point contract, we do not have the same for the reference bundler implementation.

Implementing this would allow quicker iterations when exploring new features and integrations with 4337.

## Proposed solution

Migrate the `docker-compose.yaml` file from the https://github.com/5afe/playground-4337 repository and create a dedicated E2E test that verifies support.

## Additional Context

It should also be possible to extend the setup to additional open source bundlers (such as Alchemy `rundler`, Pimlico `alto-bundler`).

### PR DESCRIPTION
**Pull request by nlordell**
Fixes #145

This PR adds an E2E test that interacts with the reference bundler implementation to ensure that the ERC-4337 module does not break any rules during user operation validation.

In addition, some documentation on how to run the E2E tests was added as well as a new CI step for executing the E2E tests.

### CODE REVIEW

File: .github/workflows/ci_4337.yml [hunk 1]
```diff
@@ -14,14 +14,41 @@ jobs:
...
+      - uses: actions/setup-node@v3
+        with:
+          node-version: 18.x
+          cache: npm
+          cache-dependency-path: 4337/package-lock.json
+      - run: npm ci
+      - run: |
+          docker compose up -d
+          # wait for containers to start up
+          SECONDS=0
```
Comment (mmv08) [inline .github/workflows/ci_4337.yml:41]: TIL https://unix.stackexchange.com/a/646605

File: 4337/docker/bundler/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,8 @@
+FROM docker.io/library/node:18
```
Comment (mmv08) [inline 4337/docker/bundler/Dockerfile:1]: The latest LTS is 20. Would it make sense to use the latest version? Or is it the bundler's requirement?

File: 4337/src/utils/safe.ts [hunk 3]
```diff
@@ -116,6 +115,10 @@ export class MultiProvider4337 extends JsonRpcProvider {
       return this.generalProvider.send(method, params)
     }
   }
+
+  public async sendUserOperation(userOp: UserOperation, entryPoint: AddressLike): Promise<string> {
+    return await super.send('eth_sendUserOperation', [userOp, await ethers.resolveAddress(entryPoint, this)])
```
Comment (mmv08) [inline 4337/src/utils/safe.ts:120]: even resolving ens names here. attention to detail 🥇

File: .github/workflows/ci_4337.yml (hunk 1 above)
Comment (mmv08) [reply .github/workflows/ci_4337.yml:41]: It doesn't seem like the timeout works, though. It took longer than 30 seconds for cold boot on my local machine, and according to the CI logs, it also takes more than 30 seconds.

File: 4337/docker/bundler/Dockerfile (hunk 2 above)
Comment (nlordell) [reply 4337/docker/bundler/Dockerfile:1]: It is what they use in their CI: https://github.com/eth-infinitism/bundler/blob/436f12241b06397989fd7a522af0deca0d46ac09/.github/workflows/build.yml#L22

File: .github/workflows/ci_4337.yml (hunk 1 above)
Comment (nlordell) [reply .github/workflows/ci_4337.yml:41]: This timeout only starts after the `docker componse up -d` command ends (I reset the counter to 0 here).

Building the docker image usually is what takes the longest (~1minute), but it isn’t counted in the timeout.

Comment (mmv08) [reply .github/workflows/ci_4337.yml:41]: Ahh it's for the start only, sorry I misunderstood it

**Review by mmv08 — APPROVED** [review]

File: 4337/docker/bundler/Dockerfile (hunk 2 above)
Comment (mmv08) [reply 4337/docker/bundler/Dockerfile:1]: > Node.js 18 will be supported until April 2025

Ok we have a plenty of time

**Pull request merged**